chainerx_cc/chainerx/routines/pooling.h

```diff
@@ -1,13 +1,45 @@
 #pragma once
 
 #include <cstdint>
+#include <memory>
+#include <tuple>
+
+#include <nonstd/optional.hpp>
 
 #include "chainerx/array.h"
 #include "chainerx/constant.h"
+#include "chainerx/op.h"
 #include "chainerx/stack_vector.h"
 
 namespace chainerx {
 
+class MaxPoolState {
+    virtual ~MacPoolState() = default;
+};
+
+class MaxPoolOp : public Op {
+public:
+    static const char* name() { return "MaxPool"; }
+
+    virtual std::tuple<Array, std::shared_ptr<MaxPoolState>> Call(
+            const Array& x, bool return_state, const nonstd::optional<Array>& out) = 0;
+};
+
+class MaxPoolGradOp : public Op {
+public:
+    static const char* name() { return "MaxPoolGrad"; }
+
+    virtual std::tuple<Array, std::shared_ptr<MaxPoolState>> Call(
+            const Array& gout, const std::shared_ptr<MaxPoolState>& state, bool return_state, const nonstd::optional<Array>& gx) = 0;
```

Comment:
This looks a bit obfuscated (but it's consistent with how the state is returned and how the logic is controlled via a boolean argument)? 🤔

Reply:
Ummm will `using shared_state = std::shared_ptr<MaxPoolState>` be nicer?

Reply:
That might help a bit but I was also concerned with the actual return and input values of the declaration.  However, they are probably all needed.